From d7b6042d9081cf89e41023518a92823bd10724bc Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Tommi=20My=C3=B6h=C3=A4nen?= <ext-tommi.1.myohanen@nokia.com>
Date: Mon, 2 Nov 2009 09:54:55 +0200
Subject: [PATCH] camerabin: send video preview when idle

Separate video preview conversion and sending operation from user-stop
signal handler to g_idle_add callback. This speeds up video stopping in
critical situations.
---
 gst/camerabin/gstcamerabin.c |   29 +++++++++++++++++++++++------
 1 files changed, 23 insertions(+), 6 deletions(-)

diff --git a/gst/camerabin/gstcamerabin.c b/gst/camerabin/gstcamerabin.c
index 1f19f9a..f5bbe22 100644
--- a/gst/camerabin/gstcamerabin.c
+++ b/gst/camerabin/gstcamerabin.c
@@ -2036,6 +2036,27 @@ gst_camerabin_reset_to_view_finder (GstCameraBin * camera,
 }
 
 /*
+ * gst_camerabin_send_video_preview:
+ * u_data: camerabin object
+ *
+ * Send video preview message to application.
+ */
+static gboolean
+gst_camerabin_send_video_preview (gpointer u_data)
+{
+  GstCameraBin *camera = GST_CAMERABIN (u_data);
+
+  if (camera->video_preview_caps && camera->video_preview_buffer) {
+    gst_camerabin_send_preview (camera, camera->video_preview_buffer);
+    gst_buffer_unref (camera->video_preview_buffer);
+    camera->video_preview_buffer = NULL;
+  }
+
+  /* disconnect automatically */
+  return FALSE;
+}
+
+/*
  * gst_camerabin_do_stop:
  * @camera: camerabin object
  *
@@ -2051,12 +2072,6 @@ gst_camerabin_do_stop (GstCameraBin * camera)
     GST_DEBUG_OBJECT (camera, "mark stop");
     camera->stop_requested = TRUE;
 
-    if (camera->video_preview_caps && camera->video_preview_buffer) {
-      gst_camerabin_send_preview (camera, camera->video_preview_buffer);
-      gst_buffer_unref (camera->video_preview_buffer);
-      camera->video_preview_buffer = NULL;
-    }
-
     /* Take special care when stopping paused video capture */
     if ((camera->active_bin == camera->vidbin) && camera->paused) {
       /* Send eos event to video bin before setting it to playing */
@@ -2072,6 +2087,8 @@ gst_camerabin_do_stop (GstCameraBin * camera)
     GST_DEBUG_OBJECT (camera, "capturing finished");
   }
   g_mutex_unlock (camera->capture_mutex);
+
+  g_idle_add (gst_camerabin_send_video_preview, camera);
 }
 
 /*
